### PR TITLE
Change consensus timespan strategy

### PR DESCRIPTION
**Pull request by Qiao-Jin**
This PR aims to provide agiler as well as smooth block time span strategy.
Block time span will be adjusted automatically according to the time consumed in last consensus.

### CODE REVIEW

**Comment by Tommo-L** [comment]
> This PR aims to provide agiler as well as smooth block time span strategy.

This PR aims to make `TimePerBlock` be the real block time.

File: src/neo/Consensus/ConsensusService.cs [hunk 1]
```diff
@@ -174,11 +177,17 @@ private void InitializeConsensus(byte viewNumber)
                 }
                 else
                 {
+                    if (block_received_gap > Blockchain.TimePerBlock / 3 && block_received_gap < Blockchain.TimePerBlock * 2)
```
Comment (shargon) [inline src/neo/Consensus/ConsensusService.cs:180]: Could you add some comments?

File: src/neo/Consensus/ConsensusService.cs [hunk 2]
```diff
@@ -174,11 +177,20 @@ private void InitializeConsensus(byte viewNumber)
                 }
                 else
                 {
+                    // Too long (if changeview happens during consensus) or too short (which will occur during synchronization) 
+                    // consensus period will be filtered. Only normal consensus should be taken into account.
+                    if (block_received_gap > Blockchain.TimePerBlock / 3 && block_received_gap < Blockchain.TimePerBlock * 2)
+                    {
+                        // Renew time per block by last consensus period.
+                        block_span = (8 * block_span + 3 * (2 * Blockchain.TimePerBlock - block_received_gap)) / 11.0;
+                    }
```
Comment (shargon) [inline src/neo/Consensus/ConsensusService.cs:186]: else block_span = Blockchain.TimePerBlock; ?

I think that we should be able to remove the variable https://github.com/neo-project/neo/pull/1593/files#diff-98ea4c9a410746331ba188dc2f45afadR32 and compute here according to block_received_gap

**Comment by shargon** [comment]
image

http://neocharts.red4sec.com/viewer/index.html?u=../1589334871/report.json

It seems that it doesn't fix the delay.

File: src/neo/Consensus/ConsensusService.cs [hunk 3]
```diff
@@ -174,11 +177,20 @@ private void InitializeConsensus(byte viewNumber)
                 }
                 else
                 {
+                    // Too long (if changeview happens during consensus) or too short (which will occur during synchronization) 
+                    // consensus period will be filtered. Only normal consensus should be taken into account.
+                    if (block_received_gap > Blockchain.TimePerBlock / 3 && block_received_gap < Blockchain.TimePerBlock * 2)
```
Comment (shargon) [inline src/neo/Consensus/ConsensusService.cs:182]: I think that we should take the average of the last X blocks, because otherwise, we only fix big delays

Comment (Tommo-L) [reply src/neo/Consensus/ConsensusService.cs:182]: Maybe we can have a try by using the last X blocks, currently we are using [EMA(exponential moving average)](https://en.wikipedia.org/wiki/Moving_average#Exponential_moving_average).

**Comment by Tommo-L** [comment]
@shargon  Could you describe the environment? Is there a lot of traffic and a small amount of bandwidth in the network? The theoretical dynamic balance value of EMA is 15 seconds, and I have seen that the other work well and acceptable.

image

**Comment by shargon** [comment]
> @shargon Could you describe the environment? Is there a lot of traffic and a small amount of bandwidth in the network? The theoretical dynamic balance value of EMA is 15 seconds, and I have seen that the other work well and acceptable.

Your snapshot (bad-day) has this configuration https://github.com/Red4Sec/neo-resilience/blob/2a24b96dd3f63a350b6e00c00f780483c03bbbca/tests/default-tests.json#L151-L200

It produce loss and corruptions between the nodes

**Comment by Qiao-Jin** [comment]
@shargon @vncoelho We have updated the algorithm and this time it should works:)

**Comment by shargon** [comment]
We will test it again

**Comment by shargon** [comment]
http://neocharts.red4sec.com/viewer/index.html?u=../1589499550/report.json

If you check this results, after a block of 17 seconds , it doesn't reduce the time for compensate the delay.

image

File: src/neo/Consensus/ConsensusService.cs (hunk 2 above)
Comment (Qiao-Jin) [reply src/neo/Consensus/ConsensusService.cs:188]: If we use `block_span = Blockchain.TimePerBlock;` we will not be able to adjust time span. We use variable `block_span` to compensate dynamically if last `block_received_gap` exceeds or doesn't reach `Blockchain.TimePerBlock`

**Comment by Qiao-Jin** [comment]
> http://neocharts.red4sec.com/viewer/index.html?u=../1589499550/report.json
> 
> If you check this results, after a block of 17 seconds , it doesn't reduce the time for compensate the delay.
> 
> image

I tried in my local env but cannot reproduce such scenario. I wonder if such occasion is because the network is so poor that consensus + persist exceed 15 seconds and variable `block_span` is already zero. If so consensus time period must be always more than 15 seconds.

**Comment by shargon** [comment]
If the last block time was 16 seconds, we should produce a new one in 14, in order to fit the 15 seconds average per block.

**Comment by Qiao-Jin** [comment]
> If the last block time was 16 seconds, we should produce a new one in 14, in order to fit the 15 seconds average per block.

Intuitively it's OK, however if network & transaction pressure keep changing this would probably lead to violent fluctuation

**Pull request closed without merging**